### PR TITLE
Switch to pybind11

### PR DESCRIPTION
**Pull request by simleo**
Reimplements the extension module with pybind11. This brings the C++ code to 127 lines, down from about 600.

The idea was to replicate the exact functionality of the extension module while leaving all Python code unchanged, including tests. The only deviation from this is that I've removed the custom `ome_files.Error` exception in favor of pybind11's automatic exception translation. This makes the C++ code much simpler and there is also an improvement on the Python side where the underlying exception is more significant than a generic error, see for instance the translation to `IndexError` in `get_index`.

Another minor improvement is that now we're supporting both signatures for `is_interleaved`.

Finally, I've exposed the upstream version to address #21:

```
>>> import ome_files
>>> ome_files.upstream_version_info
Version(major=0L, minor=3L, patch=2L, extra=u'-DEV-df1a632')
>>> ome_files.upstream_version
'0.3.2-DEV-df1a632'
```

**Testing:**
* check that the Travis build stays green and that there are no major crimes in the new C++ code
* check that that the superbuild (together with https://github.com/ome/ome-cmake-superbuild/pull/155) does not break

### CODE REVIEW

File: src/omefiles.cpp [hunk 1]
```diff
@@ -26,67 +26,139 @@
...
+        );
+	r.setMetadataStore(store);
+      })
+    .def("set_id", [](OMETIFFReader &r, std::string id) {
+	r.setId(id);
+      }, "Set the current file name.")
+    .def("set_series", [](OMETIFFReader &r, size_t series) {
+        r.setSeries(series);
+      }, "Set the active series.")
+    .def("get_series", [](OMETIFFReader &r) {
```
Comment (rleigh-codelibre) [inline src/omefiles.cpp]: The getters and non-modifying methods could probably use `const OMETIFFReader &r`?

File: src/omefiles.cpp [hunk 2]
```diff
@@ -26,67 +26,139 @@
...
+    .def("get_rgb_channel_count", [](OMETIFFReader &r, size_t channel) {
+	return r.getRGBChannelCount(channel);
+      }, "Get the number of sub-channels for the given channel.")
+    .def("is_interleaved",
+	 (bool (OMETIFFReader::*)(size_t) const) &OMETIFFReader::isInterleaved,
+	 "Whether or not the given channel is interleaved")
+    .def("is_interleaved",
+	 (bool (OMETIFFReader::*)() const) &OMETIFFReader::isInterleaved,
+	 "Whether or not the channels are interleaved")
+    .def("get_pixel_type", [](OMETIFFReader &r) {
```
Comment (rleigh-codelibre) [inline src/omefiles.cpp]: I wonder if we should expose both the OME Model enumerations and the NumPy names?  Could we add a PixelType<->NumPy name mapping in PixelProperties?

File: src/omefiles.cpp [hunk 3]
```diff
@@ -26,67 +26,139 @@
...
+	ome::files::VariantPixelBuffer buf;
+	r.openBytes(plane, buf);
+	return py::bytes(std::string(
+          reinterpret_cast<char*>(buf.data()),
+          buf.num_elements() * ome::files::bytesPerPixel(buf.pixelType())
+	));
+      }, "Obtain the image plane for the given index.")
+    .def("get_used_files", [](OMETIFFReader &r, bool noPixels = false) {
+	std::vector<std::string> fnames;
+	for (const auto &f : r.getUsedFiles(noPixels)) {
```
Comment (rleigh-codelibre) [inline src/omefiles.cpp]: Discussed in person, but mentioning here for anyone interested.  An alternative here would be to use `std::transform` with a `back_inserter` and operate on the whole container without looping (visibly).  But it adds little additional clarity compared with a range-based for loop, and so what is here is fine.

**Comment by rleigh-codelibre** [comment]
Some suggestions added, but they are only suggestions and don't need doing if not needed for this PR.  Maybe just the const references (assuming they work OK with pybind11).

Overall this is really excellent, and shows how powerful the pybind11 bindings are.  It's a great improvement, and while it's complex it's pretty readable once you understand the pattern of how it works, and it's certainly more maintainable than the previous way.

File: src/omefiles.cpp (hunk 2 above)
Comment (simleo) [inline src/omefiles.cpp]: We could do that in a future PR, better not change the Python interface too much in this one.

**Comment by rleigh-codelibre** [comment]
The last commit looks great, thanks.

**Comment by rleigh-codelibre** [comment]
The latest change looks fine to me, and the build remains green.

**Comment by rleigh-codelibre** [comment]
The include guard pragmas look fine to me.

**Comment by sbesson** [comment]
Discussed with @simleo, the new strategy looks very reasonable and much more scalable as highlighted by @rleigh-codelibre in his review. Merging to finish the investigation including the model classes wrapping.

**Pull request merged**